log.go

```diff
@@ -20,9 +20,38 @@
 
 package fx
 
-import "go.uber.org/fx/internal/fxlog"
+import (
+	"errors"
 
-// WithLogger exposes logger option for tests.
-func WithLogger(l fxlog.Logger) Option {
-	return withLogger(l)
+	"go.uber.org/fx/internal/fxlog"
+)
+
+// logBuffer will buffer all messages until a logger has been
+// initialized.
+type logBuffer struct {
+	entries []fxlog.Entry
+	logger  fxlog.Logger
+}
+
+// Log buffers or logs an entry.
+func (l *logBuffer) Log(entry fxlog.Entry) {
+	if l.logger == nil {
+		l.entries = append(l.entries, entry)
+	} else {
+		l.logger.Log(entry)
+	}
+}
+
+// Connect flushes out all buffered entries to a logger and resets them.
+func (l *logBuffer) Connect(logger fxlog.Logger) error {
+	if logger == nil {
+		return errors.New("missing a logger to connect to")
+	}
```

Comment:
We should remove this check. This check is the sole reason that Connect has to return an error, and that's making all the logic above significantly more complicated. If we make `Connect` non-error-able, the log-construction logic gets simplified signfiicantly.

We don't really need to handle the "no logger" case here. This isn't an API external users can call. We have control over it. We can assume that we'l always call it correctly, or make necessary checks in the callers.